rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,234 @@
+---
+authors: Anton Miniailo (anton@goteleport.com)
+state: draft
+---
+
+# RFD 135 - Kubernetes App discovery
+
+## Required Approvers
+
+- Engineering: `@r0mant` && `@tigrato`
+- Product: `@klizhentas` || `@xinding33`
+- Security: `@reedloden` || `@jentfoo`
+
+## What
+
+Proposes discovery of apps that are hosted inside Kubernetes clusters and exposed as services.
+
+Related issue: [#25538](https://github.com/gravitational/teleport/issues/25538)
+
+## Why
+
+Some users dynamically deploy apps into their Kubernetes clusters and expose those as Kubernetes services. When they actively
+create/update/remove apps, manually trying to mirror these changes to Teleport may be time-consuming and prone to
+errors, especially for organizations that have a large number of services running in their Kubernetes clusters.
+Teleport will be able to discover Kubernetes service and automatically register them as
+Teleport apps and keep them in up to date state, simplifying process and making user experience much better.
+
+## Scope
+
+In this document we focus on dynamically creating Teleport apps based on discovery of services in Kubernetes clusters that are running
+Teleport agents. Discovery service and app service should be running inside the Kubernetes cluster whose services are being exposed.
+
+## Details
+
+Kubernetes app discovery will build upon existing infrastructure for other discovery capabilities that are already there. We will add another
+type of matchers to the discovery service configuration - `kubernetes`. If these matchers are present, every 5 minutes discovery service will 
+poll Kubernetes cluster, inside which it is running, for a list of services. Based on this list, it will form a list of Teleport apps resources
```

Comment:
> every 5 min the service will poll

Why polling? Wouldn't setting up a controller watching events provide a better UX with near-instant propagation?

Reply:
IIRC, Kube api server closes watchers regularly so you need to reestablish the watcher with the latest resource version you received.

I just feel that it's a bit more work than what we need.

Reply:
Plus our infrastructure for discovery is currently already oriented to polling (interval is set on generic resources watcher that passes new resources to reconciler).